### ISSUE TITLE
Switch to deterministic test IDs

### ISSUE THREAD

**Issue by rprouse** [nunit/nunit#2771]
Our current test IDs use a simplistic incrementing integer, so tests are given an ID based on the order they are discovered. Between test runs, the test IDs change if new tests are added or if data driven tests return a different number of tests. It would be good if test IDs were stable and predictable between runs, even if tests are added or deleted.

**Comment by jnm2** [comment]
Should these be GUIDs? We could combine the assembly name and IL method metadata ID of each test method with the assembly names and metadata IDs of any generic type arguments along the way, and also with the hash codes of the fixture and method parameter values if any.
If people use custom reference types for fixture and method parameters, they would have the burden of implementing GetHashCode correctly. (Otherwise every test discovery would appear to have unrelated brand new test cases.)

**Comment by rprouse** [comment]
Guids aren't reproducible, but the rest is what I was thinking if that is what you meant by guid

**Comment by jnm2** [comment]
Yes, I just meant to ask, what's the format of the end result? Arbitrary string, or integer, or guid, or hex, etc?

**Comment by CharliePoole** [comment]
This seems to duplicate #1336, which already has some good discussion that should be kept. It's especially important since it involves a different third-party runner (from VS) and we clearly don't want to hastily drive this in a direction that is only useful to one partner.

IIRC, the draft spec from Microsoft says we should use a GUID (format) in one place, but allows it to be something different in another place. That was going to be one of my comments to them.

**Comment by CharliePoole** [comment]
@rprouse I realize the term "deterministic" came from Microsoft, but I think we should either define it or qualify it. Technically, our current approach is also deterministic but not necessarily predictable, which is a different matter. Also, it is stable across runs until the assembly is changed and reloaded. If we had what I call dynamic tests, it would no longer be stable between runs, but we don't have them.

My point would be that we have to specify the exact conditions under which the ids should be deterministic, predictable and stable and that we really need to do that before writing any code. In #1336, for example, @remcomulder raises the question of different versions of the same assembly, i.e. should the id be stable in that case?

**Comment by jnm2** [comment]
Hearing from Remco and others is a great idea.
It sounded like the spec intended the IDs to be stable across source changes and rebuilds and that makes good sense to me. You don't want the last known outcome of all tests to be thrown out when you start typing and rebuilding.

My guess would be that assembly name (not file name) should be hashed in. If you change the csproj's `<AssemblyName>`, you'd get a brand new series of IDs, and if you have the same namespace and type name in two different projects, there wouldn't be an ID collision.

**Comment by ChrisMaddock** [comment]
One thought - does the NUnit engine still modify the test id per assembly, to avoid collisions? Or was that just for the nunit2 addin, my memory's failing.

This would also allow us to simplify that process. 😄

**Comment by CharliePoole** [comment]
The engine provides each driver (maps to assembly) with a prefix for use in the creation of the ids. If none is provided, then of course the framework will produce ambiguous prefixes, but the engine always provides it and relies on the driver to use it.

The "prefix" doesn't have to be used literally as a prefix, although that's what is now done. It could be used as a seed of some kind when creating the guid. One issue, of course, is that the engine can potentially run tests for various frameworks. In that case, the final result is going to be up to each framework.

In the case of the adapter, there is no issue with multiple frameworks, of course, NUnit's adapter will only handle the NUnit 3 framework. However, multiple adapters could be present in a single run. I think that means it's really up to the Test Explorer to discriminate between different assemblies that may contain hte same method name.
